libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt

```diff
@@ -79,26 +75,31 @@ internal class MapboxTripSession(
 
     override suspend fun setRoutes(
         routes: List<NavigationRoute>,
-        setRoutesInfo: SetRoutesInfo,
+        setRoutes: SetRoutes,
     ): NativeSetRouteResult {
-        logD(
-            "routes update (reason: ${setRoutesInfo.reason}, " +
-                "route IDs: ${routes.map { it.id }}) - starting",
-            LOG_CATEGORY
-        )
-        val result = when (setRoutesInfo) {
-            is BasicSetRoutesInfo -> {
-                isUpdatingRoute = true
-                setRouteToNativeNavigator(routes, setRoutesInfo)
-                    .also { isUpdatingRoute = false }
+        fun logMessage(suffix: String): () -> String = {
+            "routes update (reason: ${setRoutes::class.java.simpleName}, " +
+                "route IDs: ${routes.map { it.id }}) $suffix"
+        }
+        logD(LOG_CATEGORY, logMessage("- starting"))
+
+        val result = when (setRoutes) {
```

Comment:
Here you have another `when` to get initialLegIndex, although you've introduced an extension method for this purpose.

Reply:
`when` is needed not only for the leg index but also to resolve `reason`. Also, if you have a look at L96-137. It's not just getting the leg index.

Reply:
I understand that it's not just about the index. But it's about it as well. And as a result you have 2 places where the logic of getting leg index is implemented.

Reply:
discussed: move `SetRoutes.initialLegIndex()` to `SetRoutesEx` and cut the `NAVAND-910` to address `DirectionsSession#initialLegIndex` concerns